### PR TITLE
feat: add errorformatter

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #<issue>

## ✅ Checklist

- [ ] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [ ] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

_[Short description of what has changed]_

---

## Screenshots

_[Screenshots]_

💯

### CODE REVIEW

**Review by nexxeln — COMMENTED** [review]
Love it, we should add docs for this too before merging

**Comment by juliusmarminge** [comment]
> Love it, we should add docs for this too before merging

Good shout! What do you think would be the best docs for this?

---

## Inferring errors

By default, tRPC sets up an [error formatter]() that let's you infer your Zod Errors if you get validation errors on the backend.

Example usage:

```
function MyComponent() {
  const mutation = api.post.create.useMutation();

  return (
    <form>
      <input name="title" />
      {mutation.error?.data?.zodError?.fieldErrors.title && (
        <span className="mb-8 text-red-500">
          {mutation.error.data.zodError.fieldErrors.title}
        </span>
      )}
    </form>
  );
}
```

---

Or do you have a better idea?

**Comment by nexxeln** [comment]
Looks good to me

**Comment by c-ehrlich** [comment]
> > Love it, we should add docs for this too before merging
> 
> Good shout! What do you think would be the best docs for this?
> ## Inferring errors
> 
> By default, tRPC sets up an error formatter that let's you infer your Zod Errors if you get validation errors on the backend.
> 
> Example usage:
> 
> ```tsx
> function MyComponent() {
>   const mutation = api.post.create.useMutation();
> 
>   return (
>     <form>
>       <input name="title" />
>       {mutation.error?.data?.zodError?.fieldErrors.title ? (
>         <span className="mb-8 text-red-500">
>           {mutation.error.data.zodError.fieldErrors.title}
>         </span>
>       ) : null}
> 
>       ...
>     </form>
>   );
> }
> ```
> 
> Or do you have a better idea?

docs text should be "lets" instead of "let's"
everything else looks really good.

realized while reading this pr that maybe some of the text in trpc.ts could be made a bit easier to understand overall, but thats for another pr

File: www/src/pages/en/usage/trpc.md [hunk 1]
```diff
@@ -93,6 +93,36 @@ const UserPage = () => {
 
 You'll immediately notice how good the autocompletion and typesafety is. As soon as you write `api.`, your routers will show up in autocomplete, and when you select a router, its procedures will show up as well. You'll also get a TypeScript error if your input doesn't match the validator that you defined on the backend.
 
+## Inferring errors
+
+By default, `create-t3-app` sets up an [error formatter](https://trpc.io/docs/error-formatting) that let's you infer your Zod Errors if you get validation errors on the backend.
```
Comment (c-ehrlich) [inline www/src/pages/en/usage/trpc.md:98]:
```suggestion
By default, `create-t3-app` sets up an [error formatter](https://trpc.io/docs/error-formatting) that lets you infer your Zod Errors if you get validation errors on the backend.
```

**Comment by nexxeln** [comment]
I think we can merge this now

**Comment by JClackett** [comment]
Has anyone run into the scenario where you are updating a resource and so need a data payload but also an id to actually find the resource update:

so the function would look something like this:
```ts
update: protectedProcedure.input(z.object({ id: z.string(), data: customSchemaForResource })).mutation(..
```

now when using the errorFormatter/flatten thing from zod, any errors within the data schema validator get flattened into a "data" key, and so on the frontend you dont know which field caused the error.

What are peoples practices for handling this?

**Comment by juliusmarminge** [comment]
> Has anyone run into the scenario where you are updating a resource and so need a data payload but also an id to actually find the resource update, the function would look something like this:
> 
> ```ts
> update: protectedProcedure.input(z.object({ id: z.string(), data: customSchemaForResource })).mutation(..
> ```
> 
> now when using the errorFormatter/flatten thing from zod, any errors within the data schema validator get flattened into a "data" key, and so on the frontend you dont know which field caused the error.
> 
> What are peoples practices for handling this?

What do you mean? The current formatter would return this if you send a number instead of a string to the default query on a fresh app:

```json
{
  "formErrors": [],
  "fieldErrors": {
    "text": [
      "Expected string, received number"
    ]
  }
}
```

so you have your fieldErrors with the fields that failed

**Comment by JClackett** [comment]
Just curious wow would you structure your trpc input when updating a "todo", for example? as you'll need an ID of some sort and then the actual data to update, do you just put the ID alongside all the fields that are being updated?

**Comment by juliusmarminge** [comment]
Yea probably, i had some procedure recently which posted a review on a product which was something like 

```ts
create: protectedProcedure
  .input(z.object({
    productId: z.string(),
    text: z.string().min(10),
    rating: z.number().min(0).max(5),
  })
  .mutation(...)
```

**Comment by juliusmarminge** [comment]
Ah I see...

Something like this:

```ts
export const exampleRouter = createTRPCRouter({
  hello: publicProcedure
    .input(
      z.object({
        id: z.string(),
        data: z.object({
          text: z.string().min(3),
        }),
      })
    )
    .query(({ input }) => {
      return {
        greeting: `Hello ${input.data.text}`,
      };
    }),
});
```

would return an error like this if some field within the data object is invalid:

```json
{
  "formErrors": [],
  "fieldErrors": {
    "data": [
      "String must contain at least 3 character(s)"
    ]
  }
}
```

Hmm... not sure if you can write a better error formatter to handle that form of input structure

**Comment by JClackett** [comment]
Yeah exactly, I edited my previous comment showing a workaround ill go for as I really dont wana write my own formatter :P

**Comment by juliusmarminge** [comment]
The `.format()` (instead of `.flatten()`) method on zodErrors give you something like this:

```json
{
  "_errors": [],
  "id": {
    "_errors": [
      "Expected string, received number"
    ]
  },
  "data": {
    "_errors": [],
    "text": {
      "_errors": [
        "String must contain at least 3 character(s)"
      ]
    }
  }
}
```

which you could probably work with to make it work for your usecase

**Pull request merged**